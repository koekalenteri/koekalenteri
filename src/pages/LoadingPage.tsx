import Box from '@mui/material/Box'
import CircularProgress from '@mui/material/CircularProgress'

export function LoadingPage() {
  return (
    <Box
      sx={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        flexDirection: 'column',
        minHeight: '100vh',
        backgroundColor: 'background.default',
      }}
    >
      <CircularProgress />
    </Box>
  )
}
