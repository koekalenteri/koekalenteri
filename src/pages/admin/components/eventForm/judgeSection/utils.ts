import type { DeepPartial, EventClass, EventType, Judge, PublicJudge } from '../../../../../types'
import type { PartialEvent } from '../../EventForm'

import { isSameDay } from 'date-fns'

export type PartialPublicJudge = Partial<PublicJudge>
export type PartialPublicJudgeValue = PartialPublicJudge | PartialPublicJudge[]

export const filterJudges = (
  judges: Judge[],
  eventJudges: PublicJudge[],
  id: number | undefined,
  eventType?: EventType
) =>
  judges
    .filter((j) => !eventType?.official || j.eventTypes.includes(eventType.eventType))
    .filter((j) => j.id === id || !eventJudges.find((ej) => ej.id === j.id))

export const hasJudge = (c: DeepPartial<EventClass>, id?: number) =>
  Array.isArray(c.judge) ? c.judge.find((j) => j.id === id) : c.judge?.id === id

export const filterClassesByJudgeId = (classes?: EventClass[], id?: number) => classes?.filter((c) => hasJudge(c, id))

export const toArray = (j?: PartialPublicJudge): PartialPublicJudge[] => (j ? [j] : [])
export const makeArray = (j?: PartialPublicJudgeValue) => (Array.isArray(j) ? [...j] : toArray(j))
export const selectJudge = (j?: PartialPublicJudgeValue, judge?: PublicJudge): PartialPublicJudge[] => {
  const a = makeArray(j)
  if (judge) {
    const index = a.findIndex((cj) => cj.id === judge.id)
    if (index === -1) {
      a.push(judge)
    } else {
      a[index] = judge
    }
  }
  return a
}

export const removeJudge = (j?: PartialPublicJudgeValue, id?: number): PartialPublicJudge[] => {
  const a = makeArray(j)
  return a.filter((cj) => cj.id !== id)
}

export const updateJudge = (
  event: PartialEvent,
  id: number | undefined,
  judge: PublicJudge | undefined,
  values?: DeepPartial<EventClass>[]
) => {
  const isSelected = (c: DeepPartial<EventClass>) =>
    values?.find((v) => event && isSameDay(v.date ?? event.startDate, c.date ?? event.startDate) && v.class === c.class)
  return event.classes?.map((c) => ({
    ...c,
    judge: isSelected(c) ? selectJudge(c.judge, judge) : removeJudge(c.judge, id),
  }))
}
