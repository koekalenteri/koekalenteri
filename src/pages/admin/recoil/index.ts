export * from './emailTemplates'
export * from './events'
export * from './eventTypes'
export * from './officials'
export * from './organizers'
export * from './registrations'
export * from './user'
