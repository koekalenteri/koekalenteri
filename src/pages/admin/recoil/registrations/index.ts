export * from './atoms'
// 2nd file to avoid circular deps
export * from './atoms2'
export * from './selectors'
