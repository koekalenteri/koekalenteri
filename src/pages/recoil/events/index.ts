export * from './atoms'
export * from './effects'
export * from './filters'
export * from './selectors'
export * from './types'
