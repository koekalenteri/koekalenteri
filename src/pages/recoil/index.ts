export * from './effects'
export * from './events'
export * from './judges'
export * from './registration'
export * from './user'
