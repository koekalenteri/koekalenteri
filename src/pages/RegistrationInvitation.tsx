import type { Params } from 'react-router'
import type { DogEvent, Registration } from '../types'

import { Suspense } from 'react'
import { useTranslation } from 'react-i18next'
import { Await, Navigate, useLoaderData } from 'react-router'
import Paper from '@mui/material/Paper'
import Stack from '@mui/material/Stack'
import Typography from '@mui/material/Typography'
import i18n from 'i18next'

import { getEvent } from '../api/event'
import { getRegistration, putRegistration } from '../api/registration'
import { Path } from '../routeConfig'

import LinkButton from './components/LinkButton'
import { LoadingPage } from './LoadingPage'

interface DeferredData {
  url?: string
  event: DogEvent
  registration: Registration
}

/**
 * exported for tests
 */
export const deferredLoader = async (
  id?: string,
  registrationId?: string,
  signal?: AbortSignal
): Promise<DeferredData> => {
  if (!id || !registrationId) throw new Error('invalid params')

  const [event, registration] = await Promise.all([
    getEvent(id, signal),
    getRegistration(id, registrationId, undefined, signal),
  ]).catch((e) => {
    console.log(e)
    return []
  })

  if (!event || !registration) throw new Error('not found')

  await i18n.changeLanguage(registration.language)

  if (!registration.invitationRead) {
    registration.invitationRead = true
    await putRegistration(registration, undefined, signal)
  }

  if (registration.invitationAttachment) {
    return { url: Path.invitationAttachment(registration), event, registration }
  }

  return { event, registration }
}

/**
 * @lintignore
 */
export const loader = async ({ params, request }: { params: Params<string>; request: Request }) => {
  const { id, registrationId } = params

  return {
    data: deferredLoader(id, registrationId, request.signal),
  }
}

/**
 * @lintignore
 */
export const Component = () => {
  const loaderData: { data: Promise<DeferredData> } = useLoaderData()
  const { t } = useTranslation()

  return (
    <Suspense fallback={<LoadingPage />}>
      <Await resolve={loaderData.data} errorElement={<div>{t('invitation.openingFailed')}</div>}>
        {({ url, event, registration }: DeferredData) => {
          if (!url) return <Navigate replace to={Path.registration(registration)} />

          return (
            <main style={{ padding: '8px' }}>
              <Paper sx={{ p: 1 }}>
                <Typography variant="caption">{t('invitation.event')}</Typography>
                <Typography>
                  {event.eventType} {t('dateFormat.datespan', { start: event.startDate, end: event.endDate })}{' '}
                  {event.location} ({event.name})
                </Typography>
                <Typography variant="caption">{t('invitation.registeredDog')}</Typography>
                <Typography>
                  {registration.dog.titles} {registration.dog.regNo} {registration.dog.name}
                </Typography>
                <Typography variant="caption">{t('invitation.handler')}</Typography>
                <Typography>{registration.handler.name}</Typography>
              </Paper>
              <Stack alignItems="start" sx={{ py: 1 }} gap={1} direction="row">
                <LinkButton target="_blank" to={url} text={t('invitation.open')} />
                <LinkButton download="kutsu.pdf" to={`${url}?dl`} text={t('invitation.download')} />
              </Stack>
            </main>
          )
        }}
      </Await>
    </Suspense>
  )
}
