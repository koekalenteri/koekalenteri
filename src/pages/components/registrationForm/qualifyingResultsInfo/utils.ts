import type { EventResultRequirement, EventResultRequirements, EventResultRequirementsByDate } from '../../../../rules'
import type { ManualTestResult, QualifyingResult } from '../../../../types'

import { nanoid } from 'nanoid'

import { unique } from '../../../../lib/utils'
import { objectContains } from '../validation'

const asArray = (v: EventResultRequirements | EventResultRequirement) => (Array.isArray(v) ? v : [v])

function findFirstMissing(requirements: EventResultRequirementsByDate | undefined, results: QualifyingResult[]) {
  if (!requirements || typeof requirements.rules === 'function') {
    return []
  }
  for (const rule of requirements.rules) {
    for (const opt of asArray(rule)) {
      const { count, ...rest } = opt
      if (results.filter((r) => objectContains(r, rest)).length < count) {
        return rest
      }
    }
  }
}

export function availableTypes(requirements?: EventResultRequirementsByDate, eventType?: string) {
  if (!requirements || typeof requirements.rules === 'function') {
    switch (eventType) {
      case 'NOME-B SM':
        return ['NOME-B']
      case 'NOME-A SM':
        return ['NOME-A', 'NOME-A KV']
      case 'NOWT SM':
        return ['NOWT']
    }
    return []
  }
  return unique(requirements.rules.flatMap((rule) => asArray(rule).map((opt) => opt.type)))
}

const availableSMResults = (type?: string, eventType?: string, hasKVA?: boolean) => {
  switch (eventType) {
    case 'NOME-B SM':
      return [hasKVA ? undefined : 'FI KVA-B', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOWT SM':
      return [hasKVA ? undefined : 'FI KVA-WT', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOME-A SM':
      if (type === 'NOME-A')
        return [hasKVA ? undefined : 'FI KVA-FT', 'A1 CERT', 'A1 RES-CERT', 'A1', 'A2', 'A3'].filter(Boolean)

      if (type === 'NOME-A KV') return ['EXC CACIT', 'EXC RES-CACIT', 'EXC', 'VG', 'G']
  }
  return []
}

export function availableResults(
  requirements?: EventResultRequirementsByDate,
  type?: string,
  eventType?: string,
  existingResults?: ManualTestResult[]
) {
  if (!requirements || typeof requirements.rules === 'function') {
    if (eventType?.includes(' SM')) {
      const hasKVA = existingResults?.some((r) => r.result.startsWith('FI KVA'))

      return availableSMResults(type, eventType, hasKVA)
    }
    return []
  }
  return unique(
    requirements.rules.flatMap((rule) =>
      asArray(rule)
        .filter((opt) => !type || opt.type === type)
        .map((opt) => (opt.cert ? 'CERT' : opt.result))
    )
  )
}

export function createMissingResult(
  requirements: EventResultRequirementsByDate | undefined,
  results: Array<QualifyingResult | ManualTestResult>,
  regNo: string
): ManualTestResult {
  const rule = findFirstMissing(requirements, results)
  return {
    id: nanoid(10),
    regNo,
    date: new Date(),
    official: false,
    qualifying: true,
    type: '',
    judge: '',
    location: '',
    result: '',
    class: '',
    ...rule,
  }
}

export function resultBorderColor(qualifying: boolean | undefined) {
  if (qualifying === true) {
    return 'success.light'
  }
  if (qualifying === false) {
    return 'error.main'
  }
}

export function getResultId(result: ManualTestResult | QualifyingResult) {
  if ('id' in result) {
    return result.id
  }
  return result.type + result.class + result.date.toISOString()
}
