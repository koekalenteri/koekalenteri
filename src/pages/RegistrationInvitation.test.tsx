import type { RouteObject } from 'react-router'

import { Suspense } from 'react'
import { ThemeProvider } from '@mui/material'
import { LocalizationProvider } from '@mui/x-date-pickers'
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3'
import { render, screen } from '@testing-library/react'
import { SnackbarProvider } from 'notistack'
import { RecoilRoot } from 'recoil'

import { eventWithStaticDates } from '../__mockData__/events'
import { registrationWithStaticDates } from '../__mockData__/registrations'
import theme from '../assets/Theme'
import { locales } from '../i18n'
import { Path } from '../routeConfig'
import { DataMemoryRouter, flushPromises } from '../test-utils/utils'

import { LoadingPage } from './LoadingPage'
import { Component as RegistrationInvitation } from './RegistrationInvitation'

// Mock dependencies
jest.mock('../api/event')
jest.mock('../api/registration')

describe('RegistrationInvitation', () => {
  beforeAll(() => jest.useFakeTimers())
  afterEach(() => jest.runOnlyPendingTimers())
  afterAll(() => jest.useRealTimers())

  it('renders loading indicator while loader is pending', async () => {
    const path = Path.invitation(registrationWithStaticDates)
    const routes: RouteObject[] = [
      {
        path,
        element: <RegistrationInvitation />,
        loader: async () => ({
          data: new Promise(() => {}),
        }),
        hydrateFallbackElement: <>hydrate fallback</>,
      },
    ]

    const { container } = render(
      <ThemeProvider theme={theme}>
        <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={locales.fi}>
          <RecoilRoot>
            <Suspense fallback={<LoadingPage />}>
              <SnackbarProvider>
                <DataMemoryRouter initialEntries={[path]} routes={routes} />
              </SnackbarProvider>
            </Suspense>
          </RecoilRoot>
        </LocalizationProvider>
      </ThemeProvider>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
    expect(screen.getByRole('progressbar')).toBeInTheDocument()
  })

  it('redirects to registration page when no invitation URL is available', async () => {
    const path = Path.invitation(registrationWithStaticDates)
    const routes: RouteObject[] = [
      {
        path,
        element: <RegistrationInvitation />,
        loader: async () => ({
          data: Promise.resolve({
            event: eventWithStaticDates,
            registration: registrationWithStaticDates,
          }),
        }),
        hydrateFallbackElement: <>hydrate fallback</>,
      },
      {
        path: 'r/:id/:registrationId',
        element: <>registration-page</>,
      },
    ]

    render(
      <ThemeProvider theme={theme}>
        <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={locales.fi}>
          <RecoilRoot>
            <Suspense fallback={<LoadingPage />}>
              <SnackbarProvider>
                <DataMemoryRouter initialEntries={[path]} routes={routes} />
              </SnackbarProvider>
            </Suspense>
          </RecoilRoot>
        </LocalizationProvider>
      </ThemeProvider>
    )
    await flushPromises()

    expect(screen.getByText('registration-page')).toBeInTheDocument()
  })

  it('renders invitation page when URL is available', async () => {
    const invitationUrl = '/test-invitation-url'
    const path = Path.invitation(registrationWithStaticDates)
    const routes: RouteObject[] = [
      {
        path,
        element: <RegistrationInvitation />,
        loader: async () => ({
          data: Promise.resolve({
            url: invitationUrl,
            event: eventWithStaticDates,
            registration: registrationWithStaticDates,
          }),
        }),
        hydrateFallbackElement: <>hydrate fallback</>,
      },
    ]

    const { container } = render(
      <ThemeProvider theme={theme}>
        <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={locales.fi}>
          <RecoilRoot>
            <Suspense fallback={<LoadingPage />}>
              <SnackbarProvider>
                <DataMemoryRouter initialEntries={[path]} routes={routes} />
              </SnackbarProvider>
            </Suspense>
          </RecoilRoot>
        </LocalizationProvider>
      </ThemeProvider>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()

    // Check that event information is displayed
    expect(screen.getByText(eventWithStaticDates.eventType, { exact: false })).toBeInTheDocument()

    // Check that dog information is displayed
    expect(screen.getByText(String(registrationWithStaticDates.dog.name), { exact: false })).toBeInTheDocument()

    // Check that handler information is displayed
    expect(screen.getByText(registrationWithStaticDates.handler.name)).toBeInTheDocument()

    // Check that buttons are displayed
    expect(screen.getByText('invitation.open')).toBeInTheDocument()
    expect(screen.getByText('invitation.download')).toBeInTheDocument()

    // Check that links have correct URLs
    const openButton = screen.getByText('invitation.open').closest('a')
    const downloadButton = screen.getByText('invitation.download').closest('a')

    expect(openButton).toHaveAttribute('href', invitationUrl)
    expect(downloadButton).toHaveAttribute('href', `${invitationUrl}?dl`)
    expect(downloadButton).toHaveAttribute('download', 'kutsu.pdf')
  })
})
