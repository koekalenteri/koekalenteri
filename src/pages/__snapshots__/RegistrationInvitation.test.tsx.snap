// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationInvitation renders error message when loader fails 1`] = `
<div>
  <div>
    Koekutsun avaaminen ep√§onnistui
  </div>
</div>
`;

exports[`RegistrationInvitation renders invitation page when URL is available 1`] = `
<div>
  <main
    style="padding: 8px;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-12qurmt-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
      >
        Koe
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        NOU
         
        dateFormat.datespan start, end
         
        test location
         (
        test
        )
      </p>
      <span
        class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
      >
        Ilmoitettu koira
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Dog titles
         
        TESTDOG-0010
         
        Dog Name 10
      </p>
      <span
        class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
      >
        Ohjaaja
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Handler Name
      </p>
    </div>
    <div
      class="MuiStack-root css-i74727-MuiStack-root"
    >
      <div
        class="MuiBox-root css-mro3c9"
        role="button"
      >
        <a
          class="link"
          data-discover="true"
          href="/test-invitation-url"
          target="_blank"
        >
          Avaa koekutsu
        </a>
      </div>
      <div
        class="MuiBox-root css-mro3c9"
        role="button"
      >
        <a
          class="link"
          data-discover="true"
          download="kutsu.pdf"
          href="/test-invitation-url?dl"
        >
          Lataa koekutsu
        </a>
      </div>
    </div>
  </main>
</div>
`;

exports[`RegistrationInvitation renders loading indicator while loader is pending 1`] = `
<div>
  <div
    class="MuiBox-root css-gedbwv"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1vzstpr-MuiCircularProgress-root"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</div>
`;
